### ISSUE TITLE
[projmgr] [buildmgr] Open door to more compilers

### ISSUE THREAD

**Issue by VGRSTM** [Open-CMSIS-Pack/devtools#120]
Per current schemas supported compilers are limited to AC6, IAR & GCC:
>"CompilerType": {
      "type": "string",
      "pattern": "^(GCC|AC6|IAR)(@(\\d+\\.\\d+\\.\\d+((\\+|\\-)[\\w.+-]+)?))?$",
      "description": "Selection of the toolchain used in the project (GCC, AC6, IAR), optionally with version, for example AC6@6.16.0"
    },

What about if aiming to rely on clang as example ?

**Comment by jkrech** [comment]
I think we need to start somewhere. The design should allow to extend this to other compilers going forward, however we need to identify the people contributing and maintaining the tool-chain support which currently mainly lives in CMSIS Build and the compiler specific file that live [here](https://github.com/Open-CMSIS-Pack/devtools/tree/main/tools/buildmgr/cbuildgen/config).
We are trialing this with IAR now:
see [PR116](https://github.com/Open-CMSIS-Pack/devtools/pull/116)

**Comment by VGRSTM** [comment]
@jkrech yes for sure no worries. 
May be possibly related to #68.  @brondani may have good idea about some registry requirement.

**Comment by brondani** [comment]
I believe that having the supported compiler names in the schema is a good mechanism to avoid using unsupported or misspelled compilers. Obviously it has to be kept in sync with the provided configuration files - independently of how they are registered. Currently they are part of the [Build Manager configuration](https://github.com/Open-CMSIS-Pack/devtools/tree/main/tools/buildmgr/cbuildgen/config).

**Comment by VGRSTM** [comment]
> having the supported compiler names in the schema is a good mechanism 
Yes but not so flexible / open.

Have too at a point to sync. with https://open-cmsis-pack.github.io/Open-CMSIS-Pack-Spec/main/html/pdsc_conditions_pg.html#CompilerEnumType ... link between all such is not so clear to me.

**Comment by VGRSTM** [comment]
Same question about:

> project:
  compiler: AC6                    # Use Arm Compiler 6 for this project
  device: LPC55S69JEV98:cm33_core0 # Device name (exact name as defined in the DFP) optional with Pname for core selection
  optimize: size                   # Code optimization: emphasis code size

Schema about `optimize` is quite restrictive:
>"OptimizeType": {
      "enum": [ "max", "size", "speed", "debug" ],
      "description": "Generic optimize levels (max, size, speed, debug) for code generation."
    },

AC or GCC toolchains have much more capabilities ...

PS: as addon unclear to me how such `optimize` directive is propagated to cprj ... as sounds proper place is some `cflags` which     is computed from yml `misc` contributions

**Comment by brondani** [comment]
Compiler options abstractions such as `optimize`, `debug` and `warnings` are intended for describing toolchain-independent solutions when using the most common generic configurations. If a given solution needs specialization the `misc` keyword shall be used.
The translation of abstract options into each supported toolchain specific options will be done by the Build Manager in the toolchain configuration cmake files.
Feel free to propose changes and/or extensions.

**Comment by VGRSTM** [comment]
@brondani ok got it.

Being fair I'm not fan about current toolchain configuration cmake files. Works fine but sounds to me weird to get a mixture of tooling / processing (`cbluidgen` here) and data (cmake files). According to my view toolchain data have to be contributed by some toolchain(s) contribution(s).
Sounds to me current solution is:
 + As end user: hard coded without any way to act on
 + As Open-CMSIS solution integrator: introducing highly coupled tools / toolchains solution
 
Back to compiler options abstractions such as `optimize`, `debug` and `warnings` I would prefer myself to rely on Open-CMSIS integration (first guess an IDE) to manage such abstraction valuable facilities. As integrator the way I foresee is IDE will promote to end user such options abstractions then proper writing will be done by IDE to *.yml files promoting in sync. misc. data. In other words I foresee abstraction at UI level ....

I guess I perfectly understand your first guess aim here. Adding such  compiler options abstractions such as `optimize`, `debug` and `warnings` to the standard is a way to "force" each Open-CMSIS integrator to implement such helping end user to move from one vendor to another getting same service. Sounds nice but as I've promoted Thanks other issues, I'm a bit reluctant to get a standard being too closed. As example if looking at current schema if `warnings`:
>"WarningsType": {
      "enum": [ "no", "all", "Misra", "AC5-like" ],
      "description": "Control warnings (no, all, Misra, AC5-like)."
    },
... I've feeling it perfectly matches Keil's requirements ... not sure if GCC or ... I've no issue at all about Keil's requirement at all :-)  but high level speaking each time we are introducing an `enum` we are going to face trouble.

At least my 2 cents.

**Comment by brondani** [comment]
@VGRSTM The toolchain `.cmake` files contain all the mapping that CMake needs to generate build instructions from a rather toolchain-independent CMakeLists.txt file.

Increase toolchain abstraction was one of the outcomes of the Shadowfax Componentization Exploration Team meetings with @tarek-bochkati and @DavidJurajdaNXP and it is not intended to force anyone into the standard but to provide a common and easy way of describing most embedded projects.
Indeed there is always the option to use `[c|cxx|as|ld]flags`, intended for projects that may need toolchain specialization.

The enums listed in the concept are just a proposal, it's totally open for suggestions and since it is open source it will remain open for refinements in the future. The importance of having the options in the schema is to provide a quick feedback to the user when creating/editing such fields.

**Comment by VGRSTM** [comment]
All, my proposal is move toolchain configuration discussion to dedicated issue for clarity purpose. See: https://github.com/Open-CMSIS-Pack/devtools/issues/179

Let's keep such issue to deal with free form compiler support instead enum.
If https://github.com/Open-CMSIS-Pack/devtools/blob/main/tools/projmgr/schemas/common.schema.json current proposal is:
>  "CompilerType": {
      "type": "string",
      "pattern": "^(GCC|AC6|IAR)(@(\\d+\\.\\d+\\.\\d+((\\+|\\-)[\\w.+-]+)?))?$",
      "description": "Selection of the toolchain used in the project (GCC, AC6, IAR), optionally with version, for example AC6@6.16.0"
    },
 
If https://github.com/Open-CMSIS-Pack/devtools/blob/main/tools/buildmgr/cbuildgen/config/CPRJ.xsd curren tproposal is:
>  \<!-- compiler toolchain enumeration --\>
 \<xs:simpleType name="CompilerEnumType"\>
    \<xs:restriction base="xs:token"\>
      \<xs:enumeration value="GCC" /\>
      \<xs:enumeration value="AC5" /\>
      \<xs:enumeration value="AC6" /\>
      \<xs:enumeration value="IAR" /\>
      \<xs:enumeration value="Tasking" /\>
      \<xs:enumeration value="GHS" /\>
      \<xs:enumeration value="Cosmic" /\>
      \<xs:enumeration value="G++"/\>
      \<xs:enumeration value="*" /\>
    \</xs:restriction\>
  \</xs:simpleType\>

All such in the light of pdsc we have to take care about condition resolving https://github.com/Open-CMSIS-Pack/Open-CMSIS-Pack-Spec/blob/main/schema/PACK.xsd:
>  \<!-- compiler toolchain enumeration --\>
  \<xs:simpleType name="CompilerEnumType"\>
    \<xs:restriction base="xs:token"\>
      \<xs:enumeration value="GCC" /\>
      \<xs:enumeration value="ARMCC" /\>
      \<xs:enumeration value="IAR" /\>
      \<xs:enumeration value="Tasking" /\>
      \<xs:enumeration value="GHS" /\>
      \<xs:enumeration value="Cosmic" /\>
      \<xs:enumeration value="G++"/> <!-- Code Sourcery --\>
      \<xs:enumeration value="XC"/> <!-- Microchip MPLAB XC32 --\>
      \<xs:enumeration value="*" /\>
    \</xs:restriction\>
  \</xs:simpleType\>
>
>  \<!-- compiler specific options/mode --\>
  \<xs:simpleType name="CompilerOptionsEnumType"\>
    \<xs:restriction base="xs:token"\>
      \<xs:enumeration value="AC5" /\>
      \<xs:enumeration value="AC6" /\>
      \<xs:enumeration value="AC6LTO" /\>
    \</xs:restriction\>
  \</xs:simpleType\>
  
  
Some coherency issue if yml.
I would add on top of what about if aiming to add one contribution (LLVM as example) have I to claim such is AC6 or GCC like ?

**Comment by ReinhardKeil** [comment]
Is this still relevant?

**Comment by VGRSTM** [comment]
> Is this still relevant?

To me yes. Would not claim high level priority but as part of a standard such is real problematic to me. 
As example if I want to introduce LLVM (not AC6 version)  what do I do ?
And still have https://github.com/Open-CMSIS-Pack/devtools/issues/120#issuecomment-1060396857 issue i.e. global proposal is not homogeneous.

As high level summary yes still valid but no emergency

**Comment by ReinhardKeil** [comment]
@VGRSTM agreed that we should give a high-level information how this is done.  I created therefore #493 and suggest to close this for the time being as it became a bit fuzzy.
